addon/-private/system/record-map.js

```diff
@@ -0,0 +1,47 @@
+import EmptyObject from 'ember-data/-private/system/empty-object';
+
+export default class ClassMeta {
+  constructor(modelName) {
+    this.modelName = modelName;
+    this._idToRecord = null;
+    this._records = null;
+    this._metadata = null;
+  }
+
+  get idToRecord() {
+    return this._idToRecord || (this._idToRecord = new EmptyObject());
+  }
+
+  get records() {
+    return this._records || (this._records = []);
+  }
+
+  get metadata() {
+    return this._metadata || (this._metadata = new EmptyObject());
+  }
+
+  /* deprecated way of accessing modelClass */
+  get type() {
+    throw new Error('RecordMap.type is no longer available');
+  }
+
+  clear() {
+    if (this._records) {
+      let records = this._records.slice();
```

Comment:
I'll look into it. This is the same logic preserved from before, and it's unclear if emptying this has ramifications, I think it does (I think individual records are currently responsible for removing themselves).  This is a data flow story I will untangle soon in order to also help us unlock better change tracking for record arrays.